crates/core/tedge_actors/src/servers/message_boxes.rs

```diff
@@ -80,69 +73,92 @@ impl<Request: Message, Response: Message> ConcurrentServerMessageBox<Request, Re
+    pub fn send_response_once_done(&mut self, pending_result: PendingResult) {
         self.pending_responses.push(pending_result);
     }
+}
 
-    async fn send_result(&mut self, result: Result<(usize, Response), tokio::task::JoinError>) {
-        if let Ok(response) = result {
-            let _ = self.clients.send(response).await;
-        }
-        // TODO handle error cases:
-        // - cancelled task
-        // - task panics
-        // - send fails
-    }
+/// A message box used by a client to request a server and await the responses.
+pub struct ClientMessageBox<Request, Response> {
+    sender: DynSender<RequestEnvelope<Request, Response>>,
 }
 
-/// Client side handler of requests/responses sent to an actor
-/// Client side handler that allows you to send requests to an actor
-/// and synchronously wait for its response using the `await_response` function.
-///
-/// Note that this message box sends requests and receive responses.
-pub struct ClientMessageBox<Request, Response: Debug> {
-    messages: SimpleMessageBox<Response, Request>,
+impl<Request: Message, Response: Message> Clone for ClientMessageBox<Request, Response> {
+    fn clone(&self) -> Self {
+        ClientMessageBox {
+            sender: self.sender.sender_clone(),
+        }
+    }
 }
 
 impl<Request: Message, Response: Message> ClientMessageBox<Request, Response> {
-    /// Create a new `ClientMessageBox` connected to the service.
+    /// Create a [ClientMessageBox] connected to a given [Server]
     pub fn new(
-        client_name: &str,
-        service: &mut impl ServiceProvider<Request, Response, NoConfig>,
+        server: &mut impl ServiceProvider<
+            RequestEnvelope<Request, Response>,
+            NoMessage,
+            ReplyToRequester,
+        >,
     ) -> Self {
-        let capacity = 1; // At most one response is ever expected
-        let messages = SimpleMessageBoxBuilder::new(client_name, capacity)
-            .with_connection(service)
-            .build();
-        ClientMessageBox { messages }
+        let sender = server.connect_consumer(ReplyToRequester, NullSender.into());
+        ClientMessageBox { sender }
     }
 
     /// Send the request and await for a response
     pub async fn await_response(&mut self, request: Request) -> Result<Response, ChannelError> {
-        self.messages.send(request).await?;
-        self.messages
-            .recv()
+        let (sender, receiver) = oneshot::channel::<Response>();
+        let reply_to = Box::new(Some(sender));
+        self.sender
+            .send(RequestEnvelope { request, reply_to })
+            .await?;
+        let response = receiver.await;
+        response.map_err(|_| ChannelError::ReceiveError())
+    }
+}
+
+/// A [Sender] used by a client to send requests to a server,
+/// redirecting the responses to another recipient.
+pub(crate) struct RequestSender<Request: 'static, Response: 'static> {
```

Comment:
That struct looks redundant/confusing. Was this added just for the legacy code that follows the old request/response pattern? And will be deprecated later on?

Reply:
It's not redundant.

A server actor can be used two different ways:

- By sending requests and awaiting the responses on the *same* async function (as done by the updated `c8y_firmware_manager` for download requests.
- By redirecting the responses to another message box (as done for timeout requests).

A `RequestSender` is what is returned to the client who opts in for the second pattern.

See https://github.com/thin-edge/thin-edge.io/pull/2600#discussion_r1500415792